library/core/src/any.rs

```diff
@@ -81,11 +85,76 @@
 //!     do_work(&my_i8);
 //! }
 //! ```
+//!
+//! # `Provider` and `Demand`
+//!
+//! `Provider` and the associated APIs support generic, type-driven access to data, and a mechanism
+//! for implementers to provide such data. The key parts of the interface are the `Provider`
+//! trait for objects which can provide data, and the [`request_value`] and [`request_ref`]
+//! functions for requesting data from an object which implements `Provider`. Note that end users
+//! should not call `request_*` directly, they are helper functions for intermediate implementers
+//! to use to implement a user-facing interface.
+//!
+//! Typically, a data provider is a trait object of a trait which extends `Provider`. A user will
+//! request data from the trait object by specifying the type.
+//!
+//! ## Data flow
+//!
+//! * A user requests an object, which is delegated to `request_value` or `request_ref`
+//! * `request_*` creates a `Demand` object and passes it to `Provider::provide`
+//! * The object provider's implementation of `Provider::provide` tries providing values of
+//!   different types using `Demand::provide_*`. If the type matches the type requested by
+//!   the user, it will be stored in the `Demand` object.
+//! * `request_*` unpacks the `Demand` object and returns any stored value to the user.
+//!
+//! ## Examples
+//!
+//! ```
+//! # #![allow(incomplete_features)]
+//! # #![feature(provide_any)]
+//! # #![feature(trait_upcasting)]
```

Comment:
I ended up going down a different path and making it so that the `request_ref`/`request_value` functions can take closures instead of `&dyn Provider`.

Reply:
Why do you prefer the request functions taking closures?

Reply:
I meant rather than the original suggestion here of using an impl on the client trait?

Reply:
We can change the signature of `request_value` to `pub fn request_value<'a, T: 'static, P: Provider + ?Sized>(provider: &'a P) -> Option<T>` and we get to keep the same friendly usage. Does that work?

Reply:
Without the trait upcasting. I have implemented it in my test repo and it seems to work, not sure if there is a reason to avoid it?

Reply:
> I meant rather than the original suggestion here of using an impl on the client trait?

ooh, the specific mechanism isn't important to me, so much as making it usable without trait upcasting.

> We can change the signature of `request_value` to `pub fn request_value<'a, T: 'static, P: Provider + ?Sized>(provider: &'a P) -> Option<T>` and we get to keep the same friendly usage. Does that work?

That works great!

Reply:
I've updated the PR. This would work better with https://github.com/rust-lang/rust/issues/83701, if that stabilises before this does, then we could change `P` to use `impl Trait`.